README.md
<!--
      /       |
 ___ (___  ___|
|   )|   )|   )
|__/ |  / |__/
|
-->
<p align="center">
<img src="./img/logo.png"> <br>

<a href="https://github.com/xvxx/phd/releases">
<img src="https://img.shields.io/github/v/release/xvxx/phd?include_prereleases">
</a>
<a href="https://crates.io/crates/phd">
<img src="https://img.shields.io/crates/v/phd">
</a>
<a href="https://github.com/xvxx/phd/actions?query=workflow%3Abuild">
<img src="https://github.com/xvxx/phd/workflows/build/badge.svg">
</a>
</p>

---

`phd` is a small, easy-to-use gopher server.

Point it at a directory and it'll serve up all the text files,
sub-directories, and binary files over Gopher. Any `.gph` files will
be served up as [gophermaps][map] and executable `.gph` files will be
run as a program with their output served to the client, like the
glorious cgi-bin days of yore!

### ~ special files ~

- **`header.gph`**: If it exists in a directory, its content will be
  shown above the directory's content. Put ASCII art in it.
- **`footer.gph`**: Same, but will be shown below a directory's content.
- **`index.gph`**: Completely replaces a directory's content with what's
  in this file.
- **`??.gph`**: Visiting `gopher://yoursite/1/dog/` will try to render
  `dog.gph` from disk. Visiting `/1/dog.gph` will render the raw
  content of the .gph file.
- **`.reverse`**: If this exists, the directory contents will be listed
  in reverse alphanumeric order. Useful for phloggin', if you date
  your posts.

Any line in a `.gph` file that doesn't contain tabs (`\t`) will get an
`i` automatically prefixed, turning it into a Gopher information item.

For your convenience, phd supports **[geomyidae][gmi]** syntax for
creating links:

    This is an info line.
    [1|This is a link|/help|server|port]
    [h|URL Link|URL:https://noogle.com]

`server` and `port` will get translated into the server and port of
the actively running server, eg `localhost` and `7070`.

Any line containing a tab character (`\t`) will be sent as-is to the
client, meaning you can write and serve up raw Gophermap files too.

### ~ dynamic content ~

Any `.gph` file that is marked **executable** with be run as if it
were a standalone program and its output will be sent to the client.
It will be passed three arguments: the query string (if any), the
server's hostname, and the current port. Do with them what you will.

For example:

    $ cat echo.gph
    #!/bin/sh
    echo "Hi, world! You said:" $1
    echo "1Visit Gopherpedia	/	gopherpedia.com	70"

Then:

    $ gopher-client gopher://localhost/1/echo?something
    [INFO] Hi, world! You said: something
    [LINK] Visit Gopherpedia

Or more seriously:

    $ cat figlet.gph
    #!/bin/sh
    figlet $1

then:

    $ gopher-client gopher://localhost/1/figlet?hi gopher
    [INFO]  _     _                     _
    [INFO] | |__ (_)   __ _  ___  _ __ | |__   ___ _ __
    [INFO] | '_ \| |  / _` |/ _ \| '_ \| '_ \ / _ \ '__|
    [INFO] | | | | | | (_| | (_) | |_) | | | |  __/ |
    [INFO] |_| |_|_|  \__, |\___/| .__/|_| |_|\___|_|
    [INFO]            |___/      |_|

### ~ ruby on rails ~

`sh` is fun, but for serious work you need a serious scripting
language like Ruby or PHP or Node.JS:

    $ cat sizes.gph
    #!/usr/bin/env ruby

    def filesize(file)
        (size=File.size file) > (k=1024) ? "#{size/k}K" : "#{size}B"
    end

    puts "~ file sizes ~"
    spaces = 20
    Dir[__dir__ + "/*"].each do |entry|
        name = File.basename entry
        puts "#{name}#{' ' * (spaces - name.length)}#{filesize entry}"
    end

Now you can finally share the file sizes of a directory with the world
of Gopher!

    $ phetch -r 0.0.0.0:7070/1/sizes
    i~ file sizes ~	(null)	127.0.0.1	7070
    iCargo.toml          731B	(null)	127.0.0.1	7070
    iLICENSE             1K	(null)	127.0.0.1	7070
    iMakefile            724B	(null)	127.0.0.1	7070
    itarget              288B	(null)	127.0.0.1	7070
    iphd                 248K	(null)	127.0.0.1	7070
    iCargo.lock          2K	(null)	127.0.0.1	7070
    iREADME.md           4K	(null)	127.0.0.1	7070
    img                 96B	(null)	127.0.0.1	7070
    isizes.gph           276B	(null)	127.0.0.1	7070
    isrc                 224B	(null)	127.0.0.1	7070

## ~ usage ~

    Usage:

        phd [options] <root directory>

    Options:

        -r, --render SELECTOR  Render and print SELECTOR to stdout only.
        -h, --host HOST        Hostname for links. [Default: {host}]
        -p, --port PORT        Port for links. [Default: {port}]
        -b, --bind ADDRESS     Socket address to bind to. [Default: {bind}]
        --no-color             Don't show colors in log messages.

    Other flags:

        -h, --help      Print this screen.
        -v, --version   Print phd version.

    Examples:

        phd ./path/to/site  # Serve directory over port 7070.
        phd -p 70 docs      # Serve 'docs' directory on port 70
        phd -h gopher.com   # Serve current directory over port 7070
                            # using hostname 'gopher.com'
        phd -r / ./site     # Render local gopher site to stdout.

## ~ installation ~

On macOS you can install with [Homebrew](https://brew.sh/):

    brew install xvxx/code/phd

Binaries for Linux, Mac, and Raspberry Pi are available at
gopher://phkt.io/1/releases/phd and https://github.com/xvxx/phd/releases:

- [phd-v0.1.15-linux-x86_64.tar.gz][0]
- [phd-v0.1.15-linux-armv7.tar.gz (Raspberry Pi)][1]
- [phd-v0.1.15-macos.zip][2]

Just unzip/untar the `phd` program into your `$PATH` and get going!

If you have **[cargo][rustup]**, you can install the crate directly:

    cargo install phd --locked

## ~ development ~

    cargo run -- ./path/to/gopher/site

## ~ resources ~

- gopher://bitreich.org/1/scm/geomyidae/files.gph
- https://github.com/gophernicus/gophernicus/blob/master/README.Gophermap
- https://gopher.zone/posts/how-to-gophermap/
- [rfc 1436](https://tools.ietf.org/html/rfc1436)

## ~ todo ~

- [ ] systemd config, or something
- [ ] TLS support
- [ ] user input sanitization tests

## ~ status ~

phd is no longer under active development, but the latest version works great.

[0]: https://github.com/xvxx/phd/releases/download/v0.1.15/phd-v0.1.15-linux-x86_64.tar.gz
[1]: https://github.com/xvxx/phd/releases/download/v0.1.15/phd-v0.1.15-linux-armv7.tar.gz
[2]: https://github.com/xvxx/phd/releases/download/v0.1.15/phd-v0.1.15-macos.zip
[map]: https://en.wikipedia.org/wiki/Gopher_(protocol)#Source_code_of_a_menu
[gmi]: http://r-36.net/scm/geomyidae/
[rustup]: https://rustup.rs


CHANGELOG.md
## v0.1.15

- Update `alphanumeric-sort` dependency so `cargo install phd` works again.
- Recommend `cargo install phd --locked` in the README

## v0.1.14

- If the `NO_COLOR` env variable is set, colors won't be printed to
  the log. Same as starting with `--no-color`.
  See https://no-color.org/.

## v0.1.13

- Added `--no-color` command line option to not display color when
  logging.
- Slight change to binding behavior: if `-p` is passed without `-b`,
  we'll try to bind to that port. To this easier: `phd -p 7777`
- Accept `?` as query string indicator, not just `TAB`. See #3.

## v0.1.12

`phd` now uses `-b` and `--bind` to set the host and port to
bind to. `-p` and `-h` are now strictly for URL generation.

This should hopefully make it easier to run `phd` behind a
proxy and still generate proper links.

Thanks to @bradfier for the patch!

## v0.1.11

`phd` now ships with a basic manual!

It will be installed via homebrew and (eventually) AUR.

For now you can view it by cloning the repository and running:

    man ./doc/phd.1

Enjoy!

## v0.1.10

`phd` can now render a single page to stdout, instead of starting
as a server. Those of us in the biz refer to this as "serverless".

For example, if your Gopher site lives in `/srv/gopher` and you want
to render the main page, just run:

    phd -r / /srv/gopher

This will print the raw Gopher menu to stdout!

To view the "/about" page, pass that selector:

    phd -r / /srv/gopher

Edge computing is now Gopher-scale! Enjoy!

## v0.1.9

Switch to using GitHub Actions for release automation.


src/color.rs
//! Cheesy way to easily wrap text in console colors.
//! Example:
//! ```
//! use phd::color;
//! println!("{}Error: {}{}", color::Red, "Something broke.", color::Reset);
//! ```

use std::{
    fmt,
    sync::atomic::{AtomicBool, Ordering as AtomicOrdering},
};

/// Whether to show colors or not.
/// Defaults to true.
static SHOW_COLORS: AtomicBool = AtomicBool::new(true);

/// Hide colors.
pub fn hide_colors() {
    SHOW_COLORS.swap(false, AtomicOrdering::Relaxed);
}

/// Are we showing colors are not?
pub fn showing_colors() -> bool {
    SHOW_COLORS.load(AtomicOrdering::Relaxed)
}

macro_rules! color {
    ($t:ident, $code:expr) => {
        #[allow(missing_docs)]
        pub struct $t;
        impl fmt::Display for $t {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                if showing_colors() {
                    write!(f, "\x1b[{}m", $code)
                } else {
                    write!(f, "")
                }
            }
        }
    };
}

color!(Black, 90);
color!(Red, 91);
color!(Green, 92);
color!(Yellow, 93);
color!(Blue, 94);
color!(Magenta, 95);
color!(Cyan, 96);
color!(White, 97);

color!(DarkBlack, 30);
color!(DarkRed, 31);
color!(DarkGreen, 32);
color!(DarkYellow, 33);
color!(DarkBlue, 34);
color!(DarkMagenta, 35);
color!(DarkCyan, 36);
color!(DarkWhite, 37);

color!(Reset, 0);
color!(Bold, 1);
color!(Underline, 4);


src/gopher.rs
//! Gopher type "borrowed" from phetch.
use std::fmt;

/// Gopher types are defined according to RFC 1436.
#[allow(missing_docs)]
#[derive(Copy, Clone, PartialEq, Debug)]
pub enum Type {
    Text,       // 0
    Menu,       // 1
    CSOEntity,  // 2
    Error,      // 3
    Binhex,     // 4
    DOSFile,    // 5
    UUEncoded,  // 6
    Search,     // 7
    Telnet,     // 8
    Binary,     // 9
    Mirror,     // +
    GIF,        // g
    Telnet3270, // T
    HTML,       // h
    Image,      // I
    PNG,        // p
    Info,       // i
    Sound,      // s
    Document,   // d
}

impl Type {
    /// Is this an info line?
    pub fn is_info(self) -> bool {
        self == Type::Info
    }

    /// Text document?
    pub fn is_text(self) -> bool {
        self == Type::Text
    }

    /// HTML link?
    pub fn is_html(self) -> bool {
        self == Type::HTML
    }

    /// Telnet link?
    pub fn is_telnet(self) -> bool {
        self == Type::Telnet
    }

    /// Is this a link, ie something we can navigate to or open?
    pub fn is_link(self) -> bool {
        !self.is_info()
    }

    /// Is this something we can download?
    pub fn is_download(self) -> bool {
        match self {
            Type::Binhex
            | Type::DOSFile
            | Type::UUEncoded
            | Type::Binary
            | Type::GIF
            | Type::Image
            | Type::PNG
            | Type::Sound
            | Type::Document => true,
            _ => false,
        }
    }

    /// Gopher Item Type to RFC char.
    pub fn to_char(self) -> char {
        match self {
            Type::Text => '0',
            Type::Menu => '1',
            Type::CSOEntity => '2',
            Type::Error => '3',
            Type::Binhex => '4',
            Type::DOSFile => '5',
            Type::UUEncoded => '6',
            Type::Search => '7',
            Type::Telnet => '8',
            Type::Binary => '9',
            Type::Mirror => '+',
            Type::GIF => 'g',
            Type::Telnet3270 => 'T',
            Type::HTML => 'h',
            Type::Image => 'I',
            Type::PNG => 'p',
            Type::Info => 'i',
            Type::Sound => 's',
            Type::Document => 'd',
        }
    }

    /// Create a Gopher Item Type from its RFC char code.
    pub fn from(c: char) -> Option<Type> {
        Some(match c {
            '0' => Type::Text,
            '1' => Type::Menu,
            '2' => Type::CSOEntity,
            '3' => Type::Error,
            '4' => Type::Binhex,
            '5' => Type::DOSFile,
            '6' => Type::UUEncoded,
            '7' => Type::Search,
            '8' => Type::Telnet,
            '9' => Type::Binary,
            '+' => Type::Mirror,
            'g' => Type::GIF,
            'T' => Type::Telnet3270,
            'h' => Type::HTML,
            'I' => Type::Image,
            'p' => Type::PNG,
            'i' => Type::Info,
            's' => Type::Sound,
            'd' => Type::Document,
            _ => return None,
        })
    }
}

impl fmt::Display for Type {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.to_char())
    }
}


src/server.rs
//! A simple multi-threaded Gopher server.

use crate::{color, gopher, Request, Result};
use std::{
    cmp::Ordering,
    fs::{self, DirEntry},
    io::{self, prelude::*, BufReader, Read, Write},
    net::{SocketAddr, TcpListener, TcpStream},
    os::unix::fs::PermissionsExt,
    path::Path,
    process::Command,
    str,
    sync::atomic::{AtomicBool, Ordering as AtomicOrdering},
};
use threadpool::ThreadPool;

/// phd tries to be light on resources, so we only allow a low number
/// of simultaneous connections.
const MAX_WORKERS: usize = 10;

/// how many bytes of a file to read when trying to guess binary vs text?
const MAX_PEEK_SIZE: usize = 1024;

/// Files not displayed in directory listings.
const IGNORED_FILES: [&str; 3] = ["header.gph", "footer.gph", ".reverse"];

/// Whether to print info!() messages to stdout.
/// Defaults to true.
static SHOW_INFO: AtomicBool = AtomicBool::new(true);

/// Hide info! messages.
fn hide_info() {
    SHOW_INFO.swap(false, AtomicOrdering::Relaxed);
}

/// Print status message to the server's stdout.
macro_rules! info {
    ($e:expr) => {
        if SHOW_INFO.load(AtomicOrdering::Relaxed) {
            println!("{}", $e);
        }
    };
    ($fmt:expr, $($args:expr),*) => {
        info!(format!($fmt, $($args),*));
    };
    ($fmt:expr, $($args:expr,)*) => {
        info!(format!($fmt, $($args,)*));
    };
}

/// Starts a Gopher server at the specified host, port, and root directory.
pub fn start(bind: SocketAddr, host: &str, port: u16, root: &str) -> Result<()> {
    let listener = TcpListener::bind(&bind)?;
    let full_root_path = fs::canonicalize(&root)?.to_string_lossy().to_string();
    let pool = ThreadPool::new(MAX_WORKERS);

    info!(
        "{}» Listening {}on {}{}{} at {}{}{}",
        color::Yellow,
        color::Reset,
        color::Yellow,
        bind,
        color::Reset,
        color::Blue,
        full_root_path,
        color::Reset
    );
    for stream in listener.incoming() {
        let stream = stream?;
        info!(
            "{}┌ Connection{} from {}{}",
            color::Green,
            color::Reset,
            color::Magenta,
            stream.peer_addr()?
        );
        let req = Request::from(host, port, root)?;
        pool.execute(move || {
            if let Err(e) = accept(stream, req) {
                info!("{}└ {}{}", color::Red, e, color::Reset);
            }
        });
    }
    Ok(())
}

/// Reads from the client and responds.
fn accept(mut stream: TcpStream, mut req: Request) -> Result<()> {
    let reader = BufReader::new(&stream);
    let mut lines = reader.lines();
    if let Some(Ok(line)) = lines.next() {
        info!(
            "{}│{} Client sent:\t{}{:?}{}",
            color::Green,
            color::Reset,
            color::Cyan,
            line,
            color::Reset
        );
        req.parse_request(&line);
        write_response(&mut stream, req)?;
    }
    Ok(())
}

/// Render a response to a String.
pub fn render(host: &str, port: u16, root: &str, selector: &str) -> Result<String> {
    hide_info();
    let mut req = Request::from(host, port, root)?;
    req.parse_request(&selector);
    let mut out = vec![];
    write_response(&mut out, req)?;
    Ok(String::from_utf8_lossy(&out).into())
}

/// Writes a response to a client based on a Request.
fn write_response<W>(w: &mut W, mut req: Request) -> Result<()>
where
    W: Write,
{
    let path = req.file_path();

    // check for dir.gph if we're looking for dir
    let mut gph_file = path.clone();
    gph_file.push_str(".gph");
    if fs_exists(&gph_file) {
        req.selector = req.selector.trim_end_matches('/').into();
        req.selector.push_str(".gph");
        return write_gophermap(w, req);
    } else {
        // check for index.gph if we're looking for dir
        let mut index = path.clone();
        index.push_str("/index.gph");
        if fs_exists(&index) {
            req.selector.push_str("/index.gph");
            return write_gophermap(w, req);
        }
    }

    let meta = match fs::metadata(&path) {
        Ok(meta) => meta,
        Err(_) => return write_not_found(w, req),
    };

    if path.ends_with(".gph") {
        write_gophermap(w, req)
    } else if meta.is_file() {
        write_file(w, req)
    } else if meta.is_dir() {
        write_dir(w, req)
    } else {
        Ok(())
    }
}

/// Send a directory listing (menu) to the client based on a Request.
fn write_dir<W>(w: &mut W, req: Request) -> Result<()>
where
    W: Write,
{
    let path = req.file_path();
    if !fs_exists(&path) {
        return write_not_found(w, req);
    }

    let mut header = path.clone();
    header.push_str("/header.gph");
    if fs_exists(&header) {
        let mut sel = req.selector.clone();
        sel.push_str("/header.gph");
        write_gophermap(
            w,
            Request {
                selector: sel,
                ..req.clone()
            },
        )?;
    }

    let rel_path = req.relative_file_path();

    // show directory entries
    let reverse = format!("{}/.reverse", path);
    let paths = sort_paths(&path, fs_exists(&reverse))?;
    for entry in paths {
        let file_name = entry.file_name();
        let f = file_name.to_string_lossy().to_string();
        if f.chars().nth(0) == Some('.') || IGNORED_FILES.contains(&f.as_ref()) {
            continue;
        }
        let path = format!(
            "{}/{}",
            rel_path.trim_end_matches('/'),
            file_name.to_string_lossy()
        );
        write!(
            w,
            "{}{}\t{}\t{}\t{}\r\n",
            file_type(&entry).to_char(),
            &file_name.to_string_lossy(),
            &path,
            &req.host,
            req.port,
        )?;
    }

    let footer = format!("{}/footer.gph", path.trim_end_matches('/'));
    if fs_exists(&footer) {
        let sel = format!("{}/footer.gph", req.selector);
        write_gophermap(
            w,
            Request {
                selector: sel,
                ..req.clone()
            },
        )?;
    }

    write!(w, ".\r\n");

    info!(
        "{}│{} Server reply:\t{}DIR {}{}{}",
        color::Green,
        color::Reset,
        color::Yellow,
        color::Bold,
        req.relative_file_path(),
        color::Reset,
    );
    Ok(())
}

/// Send a file to the client based on a Request.
fn write_file<W>(w: &mut W, req: Request) -> Result<()>
where
    W: Write,
{
    let path = req.file_path();
    let mut f = fs::File::open(&path)?;
    io::copy(&mut f, w)?;
    info!(
        "{}│{} Server reply:\t{}FILE {}{}{}",
        color::Green,
        color::Reset,
        color::Yellow,
        color::Bold,
        req.relative_file_path(),
        color::Reset,
    );
    Ok(())
}

/// Send a gophermap (menu) to the client based on a Request.
fn write_gophermap<W>(w: &mut W, req: Request) -> Result<()>
where
    W: Write,
{
    let path = req.file_path();

    // Run the file and use its output as content if it's executable.
    let reader = if is_executable(&path) {
        shell(&path, &[&req.query, &req.host, &req.port.to_string()])?
    } else {
        fs::read_to_string(&path)?
    };

    for line in reader.lines() {
        write!(w, "{}", gph_line_to_gopher(line, &req))?;
    }
    info!(
        "{}│{} Server reply:\t{}MAP {}{}{}",
        color::Green,
        color::Reset,
        color::Yellow,
        color::Bold,
        req.relative_file_path(),
        color::Reset,
    );
    Ok(())
}

/// Given a single line from a .gph file, convert it into a
/// Gopher-format line. Supports a basic format where lines without \t
/// get an `i` prefixed, and the geomyidae format.
fn gph_line_to_gopher(line: &str, req: &Request) -> String {
    if line.starts_with('#') {
        return "".to_string();
    }

    let mut line = line.trim_end_matches('\r').to_string();
    if line.starts_with('[') && line.ends_with(']') && line.contains('|') {
        // [1|name|sel|server|port]
        let port = req.port.to_string();
        line = line
            .replacen('|', "", 1)
            .trim_start_matches('[')
            .trim_end_matches(']')
            .replace("\\|", "__P_ESC_PIPE") // cheap hack
            .replace('|', "\t")
            .replace("__P_ESC_PIPE", "\\|")
            .replace("\tserver\t", format!("\t{}\t", req.host).as_ref())
            .replace("\tport", format!("\t{}", port).as_ref());
        let tabs = line.matches('\t').count();
        if tabs < 1 {
            line.push('\t');
            line.push_str("(null)");
        }
        // if a link is missing host + port, assume it's this server.
        // if it's just missing the port, assume port 70
        if tabs < 2 {
            line.push('\t');
            line.push_str(&req.host);
            line.push('\t');
            line.push_str(&port);
        } else if tabs < 3 {
            line.push('\t');
            line.push_str("70");
        }
    } else {
        match line.matches('\t').count() {
            0 => {
                // Always insert `i` prefix to any lines without tabs.
                line.insert(0, 'i');
                line.push_str(&format!("\t(null)\t{}\t{}", req.host, req.port))
            }
            // Auto-add host and port to lines with just a selector.
            1 => line.push_str(&format!("\t{}\t{}", req.host, req.port)),
            2 => line.push_str(&format!("\t{}", req.port)),
            _ => {}
        }
    }
    line.push_str("\r\n");
    line
}

fn write_not_found<W>(w: &mut W, req: Request) -> Result<()>
where
    W: Write,
{
    let line = format!("3Not Found: {}\t/\tnone\t70\r\n", req.selector);
    info!(
        "{}│ Not found: {}{}{}",
        color::Red,
        color::Cyan,
        req.relative_file_path(),
        color::Reset,
    );
    write!(w, "{}", line)?;
    Ok(())
}

/// Determine the gopher type for a DirEntry on disk.
fn file_type(dir: &fs::DirEntry) -> gopher::Type {
    let metadata = match dir.metadata() {
        Err(_) => return gopher::Type::Error,
        Ok(md) => md,
    };

    if metadata.is_file() {
        if let Ok(file) = fs::File::open(&dir.path()) {
            let mut buffer: Vec<u8> = vec![];
            let _ = file.take(MAX_PEEK_SIZE as u64).read_to_end(&mut buffer);
            if content_inspector::inspect(&buffer).is_binary() {
                gopher::Type::Binary
            } else {
                gopher::Type::Text
            }
        } else {
            gopher::Type::Error
        }
    } else if metadata.is_dir() {
        gopher::Type::Menu
    } else {
        gopher::Type::Error
    }
}

/// Does the file exist? Y'know.
fn fs_exists(path: &str) -> bool {
    Path::new(path).exists()
}

/// Is the file at the given path executable?
fn is_executable(path: &str) -> bool {
    if let Ok(meta) = fs::metadata(path) {
        meta.permissions().mode() & 0o111 != 0
    } else {
        false
    }
}

/// Run a script and return its output.
fn shell(path: &str, args: &[&str]) -> Result<String> {
    let output = Command::new(path).args(args).output()?;
    if output.status.success() {
        Ok(str::from_utf8(&output.stdout)?.to_string())
    } else {
        Ok(str::from_utf8(&output.stderr)?.to_string())
    }
}

/// Sort directory paths: dirs first, files 2nd, version #s respected.
fn sort_paths(dir_path: &str, reverse: bool) -> Result<Vec<DirEntry>> {
    let mut paths: Vec<_> = fs::read_dir(dir_path)?.filter_map(|r| r.ok()).collect();
    let is_dir = |entry: &fs::DirEntry| match entry.file_type() {
        Ok(t) => t.is_dir(),
        _ => false,
    };
    paths.sort_by(|a, b| {
        let a_is_dir = is_dir(a);
        let b_is_dir = is_dir(b);
        if a_is_dir && b_is_dir || !a_is_dir && !b_is_dir {
            let ord = alphanumeric_sort::compare_os_str::<&Path, &Path>(
                a.path().as_ref(),
                b.path().as_ref(),
            );
            if reverse {
                ord.reverse()
            } else {
                ord
            }
        } else if is_dir(a) {
            Ordering::Less
        } else if is_dir(b) {
            Ordering::Greater
        } else {
            Ordering::Equal // what
        }
    });
    Ok(paths)
}

#[cfg(test)]
mod tests {
    use super::*;

    macro_rules! str_path {
        ($e:expr) => {
            $e.path()
                .to_str()
                .unwrap()
                .trim_start_matches("tests/sort/")
        };
    }

    #[test]
    fn test_sort_directory() {
        let paths = sort_paths("tests/sort", false).unwrap();
        assert_eq!(str_path!(paths[0]), "zzz");
        assert_eq!(str_path!(paths[1]), "phetch-v0.1.7-linux-armv7.tar.gz");
        assert_eq!(
            str_path!(paths[paths.len() - 1]),
            "phetch-v0.1.11-macos.zip"
        );
    }

    #[test]
    fn test_rsort_directory() {
        let paths = sort_paths("tests/sort", true).unwrap();
        assert_eq!(str_path!(paths[0]), "zzz");
        assert_eq!(str_path!(paths[1]), "phetch-v0.1.11-macos.zip");
        assert_eq!(
            str_path!(paths[paths.len() - 1]),
            "phetch-v0.1.7-linux-armv7.tar.gz"
        );
    }

    #[test]
    fn test_gph_line_to_gopher() {
        let req = Request::from("localhost", 70, ".").unwrap();

        assert_eq!(
            gph_line_to_gopher("regular line test", &req),
            "iregular line test	(null)	localhost	70\r\n"
        );
        assert_eq!(
            gph_line_to_gopher("1link test	/test	localhost	70", &req),
            "1link test	/test	localhost	70\r\n"
        );

        let line = "0short link test	/test";
        assert_eq!(
            gph_line_to_gopher(line, &req),
            "0short link test	/test	localhost	70\r\n"
        );
    }

    #[test]
    fn test_gph_geomyidae() {
        let req = Request::from("localhost", 7070, ".").unwrap();

        assert_eq!(
            gph_line_to_gopher("[1|phkt.io|/|phkt.io]", &req),
            "1phkt.io	/	phkt.io	70\r\n"
        );
        assert_eq!(gph_line_to_gopher("#[1|phkt.io|/|phkt.io]", &req), "");
        assert_eq!(
            gph_line_to_gopher("[1|sdf6000|/not-real|sdf.org|6000]", &req),
            "1sdf6000	/not-real	sdf.org	6000\r\n"
        );
        assert_eq!(
            gph_line_to_gopher("[1|R-36|/]", &req),
            "1R-36	/	localhost	7070\r\n"
        );
        assert_eq!(
            gph_line_to_gopher("[1|R-36|/|server|port]", &req),
            "1R-36	/	localhost	7070\r\n"
        );
        assert_eq!(
            gph_line_to_gopher("[0|file - comment|/file.dat|server|port]", &req),
            "0file - comment	/file.dat	localhost	7070\r\n"
        );
        assert_eq!(
            gph_line_to_gopher(
                "[0|some \\| escape and [ special characters ] test|error|server|port]",
                &req
            ),
            "0some \\| escape and [ special characters ] test	error	localhost	7070\r\n"
        );
        assert_eq!(
            gph_line_to_gopher("[|empty type||server|port]", &req),
            "empty type\t\tlocalhost\t7070\r\n",
        );
    }
}


src/request.rs
//! A Request represents a Gopher request made by a client. phd can
//! serve directory listings as Gopher Menus, plain text files as
//! Text, binary files as downloads, Gophermap files as menus, or
//! executable files as dynamic content.

use crate::Result;
use std::fs;

/// This struct represents a single gopher request.
#[derive(Debug, Clone)]
pub struct Request {
    /// Gopher selector requested
    pub selector: String,
    /// Search query string, if any.
    pub query: String,
    /// Root directory of the server. Can't serve outside of this.
    pub root: String,
    /// Host of the currently running server.
    pub host: String,
    /// Port of the currently running server.
    pub port: u16,
}

impl Request {
    /// Try to create a new request state object.
    pub fn from(host: &str, port: u16, root: &str) -> Result<Request> {
        Ok(Request {
            host: host.into(),
            port,
            root: fs::canonicalize(root)?.to_string_lossy().into(),
            selector: String::new(),
            query: String::new(),
        })
    }

    /// Path to the target file on disk requested by this request.
    pub fn file_path(&self) -> String {
        format!(
            "{}/{}",
            self.root.to_string().trim_end_matches('/'),
            self.selector.replace("..", ".").trim_start_matches('/')
        )
    }

    /// Path to the target file relative to the server root.
    pub fn relative_file_path(&self) -> String {
        self.file_path().replace(&self.root, "")
    }

    /// Set selector + query based on what the client sent.
    pub fn parse_request(&mut self, line: &str) {
        self.query.clear();
        self.selector.clear();
        if let Some((i, _)) = line
            .chars()
            .enumerate()
            .find(|&(_, c)| c == '\t' || c == '?')
        {
            if line.len() > i {
                self.query.push_str(&line[i + 1..]);
                self.selector.push_str(&line[..i]);
                return;
            }
        }
        self.selector.push_str(line);

        // strip trailing /
        if let Some(last) = self.selector.chars().last() {
            if last == '/' {
                self.selector.pop();
            }
        }
    }
}


src/main.rs
use phd;
use std::process;

const DEFAULT_BIND: &str = "[::]:7070";
const DEFAULT_HOST: &str = "127.0.0.1";
const DEFAULT_PORT: u16 = 7070;

fn main() {
    let args = std::env::args().skip(1).collect::<Vec<_>>();
    let mut args = args.iter();
    let mut root = ".";
    let mut addr = DEFAULT_BIND;
    let mut host = DEFAULT_HOST;
    let mut port = DEFAULT_PORT;
    let mut render = "";

    while let Some(arg) = args.next() {
        match arg.as_ref() {
            "--version" | "-v" | "-version" => return print_version(),
            "--help" | "-help" => return print_help(),
            "--no-color" | "-no-color" => phd::color::hide_colors(),
            "--render" | "-render" | "-r" => {
                if let Some(path) = args.next() {
                    render = path;
                } else {
                    render = "/";
                }
            }
            "--bind" | "-b" | "-bind" => {
                if let Some(a) = args.next() {
                    addr = a
                }
            }
            "--port" | "-p" | "-port" => {
                if let Some(p) = args.next() {
                    port = p
                        .parse()
                        .map_err(|_| {
                            eprintln!("bad port: {}", p);
                            process::exit(1)
                        })
                        .unwrap();
                }
            }
            "-h" => {
                if let Some(h) = args.next() {
                    host = &h;
                } else {
                    return print_help();
                }
            }
            "--host" | "-host" => {
                if let Some(h) = args.next() {
                    host = &h;
                }
            }
            _ => {
                if let Some('-') = arg.chars().nth(0) {
                    eprintln!("unknown flag: {}", arg);
                    process::exit(1);
                } else {
                    root = &arg;
                }
            }
        }
    }

    // https://no-color.org/
    if std::env::var("NO_COLOR").is_ok() {
        phd::color::hide_colors()
    }

    // If port was given and socket wasn't, bind to that port.
    let bind = if port != DEFAULT_PORT && addr == DEFAULT_BIND {
        format!("[::]:{}", port).parse().unwrap()
    } else {
        addr.parse().unwrap()
    };

    if !render.is_empty() {
        return match phd::server::render(host, port, root, &render) {
            Ok(out) => print!("{}", out),
            Err(e) => eprintln!("{}", e),
        };
    }

    if let Err(e) = phd::server::start(bind, host, port, root) {
        eprintln!("{}", e);
    }
}

fn print_help() {
    println!(
        "Usage:

    phd [options] <root directory>

Options:

    -r, --render SELECTOR  Render and print SELECTOR to stdout only.
    -h, --host HOST        Hostname for links. [Default: {host}]
    -p, --port PORT        Port for links. [Default: {port}]
    -b, --bind ADDRESS     Socket address to bind to. [Default: {bind}]
    --no-color             Don't show colors in log messages.

Other flags:

    -h, --help      Print this screen.
    -v, --version   Print phd version.

Examples:

    phd ./path/to/site  # Serve directory over port 7070.
    phd -p 70 docs      # Serve 'docs' directory on port 70
    phd -h gopher.com   # Serve current directory over port 7070
                        # using hostname 'gopher.com'
    phd -r / ./site     # Render local gopher site to stdout.
",
        host = DEFAULT_HOST,
        port = DEFAULT_PORT,
        bind = DEFAULT_BIND,
    );
}

fn print_version() {
    println!("phd v{}", env!("CARGO_PKG_VERSION"));
}
